### PR TITLE
IHEGatewayV2 DQ and DR

### PR DESCRIPTION
**Pull request by jonahkaye**
Ticket: #[1667](https://github.com/metriport/metriport-internal/issues/1667)

### Dependencies

- Upstream: (https://github.com/metriport/metriport/pull/1874)

### Description

- DQs and DRs in IHE Gateway v2 

### Testing

- Local
  - [x] unit testing
- Staging
  - [x] branch to staging where full flow worked
- Production
  - [x] send signed requests to implementers with successful responses. 

### Release Plan

- [ ] Merge this

### CODE REVIEW

File: packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dq-response.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+      gateway,
+    });
+  } else {
+    const parser = new XMLParser({
+      ignoreAttributes: false,
+      attributeNamePrefix: "_",
+      textNodeName: "_text",
+      parseAttributeValue: false,
+      removeNSPrefix: true,
+    });
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dq-response.ts:127]: Updated this to allow normal TS-style property access

File: packages/core/src/external/carequality/ihe-gateway-v2/ihe-gateway-v2-logic.ts [hunk 2]
```diff
@@ -47,10 +56,112 @@ export async function createSignSendProcessXCPDRequest({
...
+    certChain,
+    publicCert,
+    privateKey,
+    privateKeyPassword,
+    patientId,
+    cxId,
+  });
+  const results = await Promise.all(
+    responses.map(async (response, index) => {
+      const outboundRequest = drRequestsGatewayV2[index];
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/ihe-gateway-v2-logic.ts:151]: fix this logic so not relying on the order of the two lists being the same

File: packages/core/src/external/carequality/ihe-gateway-v2/ihe-gateway-v2-logic.ts [hunk 3]
```diff
@@ -47,10 +56,112 @@ export async function createSignSendProcessXCPDRequest({
...
+    signedRequests,
+    certChain,
+    publicCert,
+    privateKey,
+    privateKeyPassword,
+    patientId,
+    cxId,
+  });
+  const results = responses.map((response, index) => {
+    const outboundRequest = dqRequestsGatewayV2[index];
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/ihe-gateway-v2-logic.ts:100]: fix this logic so not relying on the order of the two lists being the same

File: packages/api/src/external/carequality/document/process-outbound-document-query-resps.ts [hunk 4]
```diff
@@ -31,6 +32,7 @@ export async function processOutboundDocumentQueryResps({
   patientId,
   cxId,
   results,
+  gatewayV2,
```
Comment (jonahkaye) [inline packages/api/src/external/carequality/document/process-outbound-document-query-resps.ts:35]: flag to indicate whether to use v1 or v2 gateway

**Review by jonahkaye — COMMENTED** [review]
From an implementation perspective, this looks close to good to go. The only application level issue that is unresolved is the outbound polling lambda needs to call `getOidsWithIHEGatewayV2Enabled` to determine whether a DQ result is gatewayv1 or gatewayv2, and then set the gatewayV2 accordingly.

File: packages/lambdas/src/ihe-gateway-v2-outbound-document-query.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        typeof privateKeyPassword !== "string" ||
+        !publicCert ||
+        typeof publicCert !== "string" ||
+        !certChain ||
+        typeof certChain !== "string"
+      ) {
+        throw new Error("Failed to get secrets or one of the secrets is not a string.");
+      }
+
+      for (const request of dqRequestsGatewayV2) {
```
Comment (jonahkaye) [inline packages/lambdas/src/ihe-gateway-v2-outbound-document-query.ts:43]: do we even need this schema validation? here and everywhere else?

File: packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dq-response.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+function parseDocumentReference(
+  //eslint-disable-next-line @typescript-eslint/no-explicit-any
+  extrinsicObject: any
+): DocumentReference {
+  const slots = extrinsicObject?.Slot || [];
+  const externalIdentifiers = extrinsicObject?.ExternalIdentifier;
+  const classifications = extrinsicObject?.Classification;
+
+  const findSlotValue = (name: string): string => {
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dq-response.ts:42]: move all these to be real functions and not arrow functions

File: .husky/pre-commit [hunk 7]
```diff
@@ -5,4 +5,4 @@ npm run check-secrets
 
 npm run lint-staged
 
-npm run full-hbs-lint
+# npm run full-hbs-lint
```
Comment (Goncharo) [inline .husky/pre-commit:8]: why is this being removed?

File: packages/api/src/external/ihe-gateway-v2/ihe-gateway-v2-direct.ts [hunk 8]
```diff
@@ -1,17 +1,36 @@
...
     super();
+    this.samlCertsAndKeys = {
+      certChain: Config.getCQOrgCertificateIntermediate(),
+      publicCert: Config.getCQOrgCertificate(),
+      privateKey: Config.getCQOrgPrivateKey(),
+      privateKeyPassword: Config.getCQOrgPrivateKeyPassword(),
+    };
+    this.pdResponseUrl = Config.getApiUrl() + "/internal/carequality/patient-discovery/response";
+    this.dqResponseUrl = Config.getApiUrl() + "/internal/carequality/document-query/response";
+    this.drResponseUrl = Config.getApiUrl() + "/internal/carequality/document-retrieval/response";
```
Comment (Goncharo) [inline packages/api/src/external/ihe-gateway-v2/ihe-gateway-v2-direct.ts:31]: nit: can put `/internal/carequality` in a const and reuse

File: packages/core/src/external/aws/s3.ts [hunk 9]
```diff
@@ -217,30 +217,69 @@ export class S3Utils {
...
-          Bucket: bucket,
-          Key: key,
-          Body: file,
-        },
-        (err, data) => {
-          if (err) {
-            console.error("Error during upload:", err);
-            reject(err);
-          } else {
-            console.log("Upload successful");
```
Comment (Goncharo) [inline packages/core/src/external/aws/s3.ts:237]: this was removed, but not added back, intentional?

File: packages/core/src/external/carequality/ihe-gateway-v2/__tests__/files/test.jpeg
Comment (Goncharo) [inline packages/core/src/external/carequality/ihe-gateway-v2/__tests__/files/test.jpeg:1]: holy fuck

File: packages/core/src/external/carequality/ihe-gateway-v2/__tests__/files/test.png
Comment (Goncharo) [inline packages/core/src/external/carequality/ihe-gateway-v2/__tests__/files/test.png:1]: why is there 2 images, but this one is 1.2MB vs the other at 142kb?

File: packages/core/src/external/carequality/ihe-gateway-v2/saml-client.ts [hunk 10]
```diff
@@ -55,76 +95,215 @@ export async function sendSignedXml({
...
+      const errorString: string = errorToString(error);
+      const extra = {
+        errorString,
+        request,
+        patientId,
+        cxId,
+      };
+      capture.error(msg, {
+        extra: {
+          context: `lambda.ihe-gateway-v2-saml-client`,
```
Comment (Goncharo) [inline packages/core/src/external/carequality/ihe-gateway-v2/saml-client.ts:284]: nit: can put in const and reuse

File: packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dq-response.ts [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    repositoryUniqueId: findSlotValue("repositoryUniqueId"),
+    docUniqueId: findExternalIdentifierValue("urn:uuid:2e82c1f6-a085-4c72-9da3-8640a32e42ab"),
+    contentType: extrinsicObject?._mimeType,
+    language: findSlotValue("languageCode"),
+    size: parseInt(findSlotValue("size")),
+    title: findClassificationName("urn:uuid:41a5887f-8865-4c09-adf7-e362475b143a"),
+    creation: findSlotValue("creationTime"),
+    authorInstitution: findClassificationSlotValue(
+      "urn:uuid:93606bcf-9494-43ec-9b4e-a7748d1a838d",
+      "authorInstitution"
```
Comment (Goncharo) [inline packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dq-response.ts:95]: wtf do all these UUIDs mean?

suggest putting them in an appropriately named const for readability, and so we can re-use later if needed

File: packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dr-response.ts [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    await s3Utils.uploadFile({
+      bucket,
+      key: filePath,
+      file: decodedBytes,
+      contentType: mimeType,
+    });
+  }
+
+  return {
+    url: "https://" + bucket + ".s3." + region + ".amazonaws.com/" + filePath,
```
Comment (Goncharo) [inline packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dr-response.ts:79]: we have an existing helper for this https://github.com/metriport/metriport/blob/09611d558fb6272338d4f9e32d2c1f11823be5ab/packages/core/src/external/aws/s3.ts#L159-L161

File: packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dr-response.ts [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+    const status = jsonObj?.Envelope?.Body?.RetrieveDocumentSetResponse?.RegistryResponse?._status
+      ?.split(":")
+      .pop();
+    const registryErrorList =
+      jsonObj?.Envelope?.Body?.RetrieveDocumentSetResponse?.RegistryResponse?.RegistryErrorList;
+    const documentResponses =
+      jsonObj?.Envelope?.Body?.RetrieveDocumentSetResponse?.DocumentResponse;
+    const soapFault = jsonObj?.Envelope?.Body?.Fault;
+
+    if ((status === "Success" || status === "PartialSuccess") && documentResponses) {
```
Comment (Goncharo) [inline packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dr-response.ts:177]: we use these strings a lot - let's put em in consts in a shared location

... if they're not string literals already (ts is hard to review this way)

File: packages/core/src/util/file-type.ts [hunk 14]
```diff
@@ -42,6 +42,12 @@ const XML_MAGIC_NUMBER_3 = 0x78;
...
 const XML_MAGIC_NUMBER_5 = 0x6c;
 const XML_MAGIC_NUMBER_6 = 0x20;
+const XML_MAGIC_NUMBER_7 = 0x3c;
+const XML_MAGIC_NUMBER_8 = 0x43;
+const XML_MAGIC_NUMBER_9 = 0x6c;
+const XML_MAGIC_NUMBER_10 = 0x69;
+const XML_MAGIC_NUMBER_11 = 0x6e;
+const XML_MAGIC_NUMBER_12 = 0x69;
 const PNG_MAGIC_NUMBER_1 = 0x89;
 const PNG_MAGIC_NUMBER_2 = 0x50;
```
Comment (Goncharo) [inline packages/core/src/util/file-type.ts:52]: what's the logic behind adding this? seems unrelated to v2

File: packages/lambdas/src/ihe-gateway-v2-outbound-document-query.ts [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        typeof privateKey !== "string" ||
+        !privateKeyPassword ||
+        typeof privateKeyPassword !== "string" ||
+        !publicCert ||
+        typeof publicCert !== "string" ||
+        !certChain ||
+        typeof certChain !== "string"
+      ) {
+        throw new Error("Failed to get secrets or one of the secrets is not a string.");
+      }
```
Comment (Goncharo) [inline packages/lambdas/src/ihe-gateway-v2-outbound-document-query.ts:41]: this whole thing can be put in shared code for the lambdas

File: packages/lambdas/src/ihe-gateway-v2-outbound-document-query.ts (hunk 5 above)
Comment (Goncharo) [reply packages/lambdas/src/ihe-gateway-v2-outbound-document-query.ts:43]: why not - seems like it'd make the code more robust

File: scratch/dr/xcadr.xml
Comment (Goncharo) [inline scratch/dr/xcadr.xml:1]: remove this - is the `gitignore` not working properly?

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: packages/core/src/external/carequality/ihe-gateway-v2/__tests__/files/test.png
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/__tests__/files/test.png:1]: PNG vs JPG file format. They have different magic numbers and needed to all file types

File: packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dq-response.ts (hunk 11 above)
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/xca/process-dq-response.ts:95]: These are the UUIDs that are always used for the specific values in the DR. So for example for authorInstitution: 
https://gazelle.ihe.net/XDStarClient/rimihe-templates/rimihe-Author_XDSDocumentEntry.html

Good call on the constants though will refactor

File: packages/core/src/util/file-type.ts (hunk 14 above)
Comment (jonahkaye) [reply packages/core/src/util/file-type.ts:52]: Its just adding this edge case logic we have in the current IHEGateway code: https://github.com/metriport/metriport/blob/33dc2739a380cad94414fa3792e1e6e81b100fc2/packages/ihe-gateway/server/CodeTemplates/Common%20Library/File%20type/File%20type.js#L42

Basically some xmls start with <clinical document> as opposed to <?xml so its just accounting for that case. There are two different test xml files to reflect these two different cases

File: packages/core/src/external/aws/s3.ts [hunk 16]
```diff
@@ -217,30 +217,69 @@ export class S3Utils {
...
-            resolve(data);
-          }
+      const uploadParams: AWS.S3.PutObjectRequest = {
+        Bucket: bucket,
+        Key: key,
+        Body: file,
+      };
+
+      if (contentType) {
+        uploadParams.ContentType = contentType;
```
Comment (jonahkaye) [inline packages/core/src/external/aws/s3.ts:239]: explicitly setting the content type on uploads. This is a bug we had in gateway v1 that caused the whole cx with outdated content types on s3

File: packages/core/src/external/aws/s3.ts (hunk 9 above)
Comment (jonahkaye) [reply packages/core/src/external/aws/s3.ts:237]: You mean the specific log line: 
`console.log("Upload successful");` ? Will add back

**Comment by Goncharo** [comment]
approved conditionally on comments being addressed

File: packages/core/src/util/file-type.ts [hunk 17]
```diff
@@ -92,13 +98,13 @@ export function isLikelyTextFile(fileBuffer: Buffer): boolean {
...
- * contents of a file. The first 5 bytes of this buffer, which generally contain the magic number, are used to
+ * contents of a file. The first 6 bytes of this buffer, which generally contain the magic number, are used to
  * identify the file type.
  * @returns The function `detectFileType` returns a string representing the detected file type.
  */
-export function detectFileType(document: string): [string, string] {
+export function detectFileType(document: Buffer): { mimeType: string; extension: string } {
   const maxBytesNeeded = 6; //NOTE: if you update detectFileType, you might need to update this number
-  const fileBuffer = Buffer.from(document.slice(0, maxBytesNeeded));
+  const fileBuffer = document.slice(0, maxBytesNeeded);
```
Comment (jonahkaye) [inline packages/core/src/util/file-type.ts:107]: the logic of Bufer.from is now just used before calling this function. This function is now shared by CW and CQ.

File: packages/core/src/external/commonwell/document/document-downloader-local.ts [hunk 18]
```diff
@@ -100,22 +100,23 @@ export class DocumentDownloaderLocal extends DocumentDownloader {
     }
 
     const old_extension = path.extname(fileInfo.name);
-    const [detectedFileType, detectedExtension] = detectFileType(downloadedDocument);
+    const documentBuffer = Buffer.from(downloadedDocument);
```
Comment (jonahkaye) [inline packages/core/src/external/commonwell/document/document-downloader-local.ts:103]: just moving Buffer.from from inside detectFileType to outside it right here, so that CW and CQ can share this detectFileType function. Also moved the function to return an object instead of an array

File: packages/lambdas/src/ihe-gateway-v2-outbound-document-query.ts (hunk 5 above)
Comment (leite08) [reply packages/lambdas/src/ihe-gateway-v2-outbound-document-query.ts:20]: If we use the same type on the lambda and on the caller, I think we don't need the validation

File: packages/lambdas/src/ihe-gateway-v2-outbound-document-query.ts [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+import { DQRequestGatewayV2Params } from "@metriport/core/external/carequality/ihe-gateway-v2/ihe-gateway-v2";
+import { createSignSendProcessDQRequests } from "@metriport/core/external/carequality/ihe-gateway-v2/ihe-gateway-v2-logic";
+import { getEnvVarOrFail, getEnvType } from "@metriport/core/util/env-var";
+import { getSamlCertsAndKeys } from "./shared/secrets";
+import { capture } from "./shared/capture";
+
+const apiUrl = getEnvVarOrFail("API_URL");
+const documentQueryResponseUrl = `http://${apiUrl}/internal/carequality/document-query/response`;
+
+export const handler = Sentry.AWSLambda.wrapHandler(
```
Comment (leite08) [inline packages/lambdas/src/ihe-gateway-v2-outbound-document-query.ts:12]: ditto: missing Sentry init code

File: packages/lambdas/src/ihe-gateway-v2-outbound-document-retrieval.ts [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+import { getSamlCertsAndKeys } from "./shared/secrets";
+import { capture } from "./shared/capture";
+
+const apiUrl = getEnvVarOrFail("API_URL");
+const drResponseUrl = `http://${apiUrl}/internal/carequality/document-retrieval/response`;
+
+export const handler = Sentry.AWSLambda.wrapHandler(
+  async ({ patientId, cxId, requestId, drRequestsGatewayV2 }: DRRequestGatewayV2Params) => {
+    try {
+      console.log(
```
Comment (leite08) [inline packages/lambdas/src/ihe-gateway-v2-outbound-document-retrieval.ts:15]: ditto: out.log

File: packages/lambdas/src/ihe-gateway-v2-outbound-patient-discovery.ts [hunk 21]
```diff
@@ -1,80 +1,56 @@
...
 const apiUrl = getEnvVarOrFail("API_URL");
-const patientDiscoveryUrl = `${apiUrl}/internal/carequality/patient-discovery/response`;
+const pdResponseUrl = `http://${apiUrl}/internal/carequality/patient-discovery/response`;
 
-// get secrets
-const privateKeySecretName = Config.getCQOrgPrivateKey();
-const privateKeyPasswordSecretName = Config.getCQOrgPrivateKeyPassword();
-const publicCertSecretName = Config.getCQOrgCertificate();
-const certChainSecretName = Config.getCQOrgCertificateIntermediate();
+export const handler = Sentry.AWSLambda.wrapHandler(
```
Comment (leite08) [inline packages/lambdas/src/ihe-gateway-v2-outbound-patient-discovery.ts:12]: ditto: missing Sentry init code

File: packages/lambdas/src/ihe-gateway-v2-outbound-patient-discovery.ts [hunk 22]
```diff
@@ -1,80 +1,56 @@
...
+      );
 
-export const handler = Sentry.AWSLambda.wrapHandler(async (event: PDRequestGatewayV2Params) => {
-  try {
-    const { patientId, cxId, pdRequestGatewayV2 } = event;
+      // validate request
+      const xcpdRequest = outboundPatientDiscoveryReqSchema.safeParse(pdRequestGatewayV2);
+      if (!xcpdRequest.success) {
+        const msg = `Invalid XCPD request - Does not conform to schema.`;
+        capture.error(msg, {
```
Comment (leite08) [inline packages/lambdas/src/ihe-gateway-v2-outbound-patient-discovery.ts:24]: dup error to Sentry

File: packages/lambdas/src/shared/secrets.ts [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+export async function getSamlCertsAndKeys(): Promise<SamlCertsAndKeys> {
+  const privateKeySecretName = Config.getCQOrgPrivateKey();
+  const privateKeyPasswordSecretName = Config.getCQOrgPrivateKeyPassword();
+  const publicCertSecretName = Config.getCQOrgCertificate();
+  const certChainSecretName = Config.getCQOrgCertificateIntermediate();
+
+  const privateKey = await getSecret(privateKeySecretName);
+  const privateKeyPassword = await getSecret(privateKeyPasswordSecretName);
+  const publicCert = await getSecret(publicCertSecretName);
+  const certChain = await getSecret(certChainSecretName);
```
Comment (leite08) [inline packages/lambdas/src/shared/secrets.ts:14]: Promise.all

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/__tests__/detect-file-type.test.ts [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+
+describe("detectFileType for all file types with non-xml body flow", () => {
+  const nonXmlBodyTemplatePath = path.join(__dirname, "./files/non-xml-body.xml");
+  const nonXmlBodyTemplate = fs.readFileSync(nonXmlBodyTemplatePath, "utf8");
+
+  testFiles.forEach(({ name, mimeType, extension }) => {
+    describe(`${name}`, () => {
+      const fileContent = fs.readFileSync(path.join(__dirname, `./files/${name}`));
+      const fileContentB64 = fileContent.toString("base64");
+      const modifiedNonXmlBody = nonXmlBodyTemplate.replace(
```
Comment (jonahkaye) [inline packages/core/src/external/carequality/ihe-gateway-v2/outbound/__tests__/detect-file-type.test.ts:15]: inserting b64 into file instead of having it there from the start

File: packages/core/src/external/carequality/ihe-gateway-v2/saml/security/utils.ts
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/saml/security/utils.ts:1]: I put it in src/utils/urn.ts

File: packages/core/src/external/carequality/ihe-gateway-v2/utils.ts [hunk 25]
```diff
@@ -1,8 +1,52 @@
...
+  fileName,
+}: {
+  cxId: string;
+  patientId: string;
+  fileName: string;
+}): string {
+  return `${cxId}/${patientId}/${fileName}`;
+}
+
+export function wrapIdInUrnUuid(id: string): string {
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/utils.ts:46]: nope that one is wrapIdInUrnOid this is wrapIdInUrnUuid!

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts [hunk 26]
```diff
@@ -0,0 +1,203 @@
...
+    {}
+  );
+
+  //eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const documentReferences = Array.isArray(documentResponses)
+    ? await Promise.all(
+        documentResponses.map(async (documentResponse: DocumentResponse) =>
+          parseDocumentReference({ documentResponse, outboundRequest, s3Utils, idMapping })
+        )
+      )
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/process/dr-response.ts:126]: not sure what you mean. parseDocumentReference is async and we use Promise.all, so that is using promises

File: packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/send/dq-requests.ts [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+  const responses = await Promise.allSettled(requestPromises);
+  const processedResponses = responses
+    .map(result => {
+      if (result.status === "fulfilled") {
+        return result.value;
+      } else {
+        return undefined;
+      }
+    })
+    .filter((response): response is DQSamlClientResponse => response !== undefined);
```
[the remark this answers is not in the document]
Comment (jonahkaye) [reply packages/core/src/external/carequality/ihe-gateway-v2/outbound/xca/send/dq-requests.ts:81]: discussed over huddle. is typecasts and so moved to just using Promise.all
